plugins/trader/portfolioManager.js

```diff
@@ -137,195 +139,50 @@ Manager.prototype.setTicker = function(callback) {
-        '(' + parseFloat(amount).toFixed(8) + ' @',
-        parseFloat(price).toFixed(8),
-        ') at',
-        this.exchange.name
-      );
-    }
-
-    log.info(
-      'Attempting to BUY',
-      order.amount,
-      this.asset,
-      'at',
-      this.exchange.name,
-      'price:',
-      order.price
-    );
-
-    this.exchange.buy(order.amount, order.price, this.noteOrder);
-  }
+Manager.prototype.trade = function(what) {
 
-  if (_.has(this.exchange, 'getLotSize')) {
-    this.exchange.getLotSize('buy', amount, price, _.bind(process));
-  } else {
-    minimum = this.getMinimum(price);
-    process(undefined, { amount: amount, price: price });
-  }
-};
+  var makeNewTrade = function(){
+    this.newTrade(what)
+  }.bind(this)
 
-// first do a quick check to see whether we can sell
-// the asset, if so SELL and keep track of the order
-// (amount is in asset quantity)
-Manager.prototype.sell = function(amount, price) {
-  let minimum = 0;
-  let process = (err, order) => {
-    // if order to small
-    if (!order.amount || order.amount < minimum) {
-      return log.warn(
-        'Wanted to buy',
-        this.currency,
-        'but the amount is too small ',
-        '(' + parseFloat(amount).toFixed(8) + ' @',
-        parseFloat(price).toFixed(8),
-        ') at',
-        this.exchange.name
-      );
+  // if an active trade is currently happening
+  if(this.currentTrade && this.currentTrade.isActive){
+    if(this.currentTrade.action !== what){
+      // if the action is different, stop the current trade, then start a new one
+      this.currentTrade.stop(makeNewTrade)
+    } else{
+      // do nothing, the trade is already going
     }
-
-    log.info(
-      'Attempting to SELL',
-      order.amount,
-      this.asset,
-      'at',
-      this.exchange.name,
-      'price:',
-      order.price
-    );
-
-    this.exchange.sell(order.amount, order.price, this.noteOrder);
-  }
-
-  if (_.has(this.exchange, 'getLotSize')) {
-    this.exchange.getLotSize('sell', amount, price, _.bind(process));
   } else {
-    minimum = this.getMinimum(price);
-    process(undefined, { amount: amount, price: price });
+    makeNewTrade()
   }
 };
 
-Manager.prototype.noteOrder = function(err, order) {
-  if(err) {
-    util.die(err);
+// instantiate a new trade object
+// TODO - impliment different trade execution types / strategies
+//        by invoking variable trade subclasses
+//      - pass the trade a specific amount limited by a currency allowance
+Manager.prototype.newTrade = function(what) {
+  // push the current (asummed to be inactive) trade to the history
+  if(this.currentTrade){
+    this.tradeHistory.push(this.currentTrade)
+    this.currentTrade = false
   }
-
-  this.orders.push(order);
-
-  // If unfilled, cancel and replace order with adjusted price
-  let cancelDelay = this.conf.orderUpdateDelay || 1;
-  setTimeout(this.checkOrder, util.minToMs(cancelDelay));
+  return this.currentTrade = new Trade(this,{action: what})
```

Comment:
If we want to stick to OO: can you not pass the parent (portfolioManager into the trade), these circular references are tightly coupled.

Reply:
Yes, it's not ideal, let's fix this.

Currently there are multiple references to the manager inside the trade class, which would call for some restructuring. This is what we could do:

- A new exchange object could be created inside the Trade class so it's not referencing it in the Manager, thus removing references to `manager.exchange`
- We can pass in the current pair currency/asset handles into the Trade class on creation, and remove `manager.asset` and `manager.currency`
- Currently the following methods are being called by the Trade class, how do you suggest we handle this?
```
this.manager.setTicker
this.manager.setPortfolio
this.manager.setFee
this.manager.getMinimum
this.manager.convertPortfolio
```

Reply:
> this.manager.convertPortfolio

This is currently only needed for emitting events upstream, the trade class doesn't need it.

Some other functions won't be needed anymore by the portfolioManager, like setFee and setTicker (we will need this one later again, for the [portfolioValueChange](https://github.com/askmike/gekko/blob/0d975707009456720f59b98c102c334c05e35585/docs/internals/events.md#portfoliovaluechange-event) event for example). All the other functions will be used differently anyway, so I propose the portfolioManager passes the exchange (`this.exchange`) to the tradeclass which reimplements what it needs, we can also drop in the portfolioManager what we don't need. What do you think?

Reply:
@askmike OK that makes sense, let's keep a single Exchange instance, and move the methods into the Trade class. I'll make that happen, test it and update this PR.